flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java

```diff
@@ -292,40 +392,31 @@ public void unregisterTask(Task task) {
 		LOG.debug("Unregister task {} from network environment (state: {}).",
 				task.getTaskInfo().getTaskNameWithSubtasks(), task.getExecutionState());
 
-		final ExecutionAttemptID executionId = task.getExecutionId();
-
 		synchronized (lock) {
 			if (isShutdown) {
 				// no need to do anything when we are not operational
 				return;
 			}
 
 			if (task.isCanceledOrFailed()) {
-				resultPartitionManager.releasePartitionsProducedBy(executionId, task.getFailureCause());
+				resultPartitionManager.releasePartitionsProducedBy(task.getExecutionId(), task.getFailureCause());
 			}
 
-			for (ResultPartition partition : task.getProducedPartitions()) {
+			for (ResultPartition partition : allPartitions.get(task.getTaskInfo().getTaskId())) {
```

Comment:
If we had just execution id to index task partitions/gates, `unregisterTask` could just accept `ExecutionId` and cause in case of failure. The first log statement with task specifics could go outside, maybe. Just less dependency on Task class.

Reply:
Yes, I agree with your idea. 
Even though we remove the first log and use `executionId` for index, we still need to consider how to refactor the uses of `task.isCanceledOrFailed()` and `task.getFailureCause()` in this method. I would try to refactor it in this way.

Reply:
the method could also accept `cause` additionally to id. `cause != null` would mean failure case.